cmd/upgrade.go

```diff
@@ -153,71 +159,14 @@ func newChartCommand() *cobra.Command {
 		}, "\n"),
+		Args: func(cmd *cobra.Command, args []string) error {
+			return checkArgsLength(len(args), "release name", "chart path")
+		},
 		RunE: func(cmd *cobra.Command, args []string) error {
-			// Note that we can't just move this block to PersistentPreRunE,
-			// because cmd.SetArgs(args) does not persist between PersistentPreRunE and RunE.
-			// The choice is between:
-			// 1. Doing this in RunE
-			// 2. Doing this in PersistentPreRunE, saving args somewhere, and calling cmd.SetArgs(args) again in RunE
-			// 2 is more complicated without much benefit, so we choose 1.
-			{
-				const (
-					dryRunUsage = "--dry-run, --dry-run=client, or --dry-run=true disables cluster access and show diff as if it was install. Implies --install, --reset-values, and --disable-validation." +
-						" --dry-run=server enables the cluster access with helm-get and the lookup template function."
-				)
-
-				cmdFlags := cmd.Flags()
-
-				// see: https://github.com/databus23/helm-diff/issues/537
-				cmdFlags.ParseErrorsWhitelist.UnknownFlags = unknownFlags
-
-				legacyDryRunFlagSet := pflag.NewFlagSet("upgrade", pflag.ContinueOnError)
-				legacyDryRun := legacyDryRunFlagSet.Bool("dry-run", false, dryRunUsage)
-				if err := legacyDryRunFlagSet.Parse(args); err == nil && *legacyDryRun {
-					diff.dryRunModeSpecified = true
-					args = legacyDryRunFlagSet.Args()
-				} else {
-					cmdFlags.StringVar(&diff.dryRunMode, "dry-run", "", dryRunUsage)
-				}
-
-				// Here we parse the flags ourselves so that we can support
-				// both --dry-run and --dry-run=ARG syntaxes.
-				//
-				// If you don't do this, then cobra will treat --dry-run as
-				// a "flag needs an argument: --dry-run" error.
-				//
-				// This works becase we have `DisableFlagParsing: true`` above.
-				// Never turn that off, or you'll get the error again.
-				if err := cmdFlags.Parse(args); err != nil {
-					return err
-				}
-
-				args = cmdFlags.Args()
-
-				if !diff.dryRunModeSpecified {
-					dryRunModeSpecified := cmd.Flags().Changed("dry-run")
-					diff.dryRunModeSpecified = dryRunModeSpecified
-
-					if dryRunModeSpecified && !(diff.dryRunMode == "client" || diff.dryRunMode == "server") {
-						return fmt.Errorf("flag %q must take an argument of %q or %q but got %q", "dry-run", "client", "server", diff.dryRunMode)
-					}
-				}
-
-				cmd.SetArgs(args)
-
-				// We have to do this here because we have to sort out the
-				// --dry-run flag ambiguity to determine the args to be checked.
-				//
-				// In other words, we can't just do:
-				//
-				// cmd.Args = func(cmd *cobra.Command, args []string) error {
-				//     return checkArgsLength(len(args), "release name", "chart path")
-				// }
-				//
-				// Because it seems to take precedence over the PersistentPreRunE
-				if err := checkArgsLength(len(args), "release name", "chart path"); err != nil {
-					return err
-				}
+			if diff.dryRunMode == "" {
+				diff.dryRunMode = "none"
```

Comment:
We can indeed use either `""` or `"none"` to denote that the dry run flag is unspecified.
However I intentionally chose `"none"` to make code consistent with helm. They use `"none"` to indicate unspecified dry-run too..